os/os.proto

```diff
@@ -0,0 +1,321 @@
+// Client progresses through 3 RPCs:
+//   1) Installation - provide the Target with the OS package.
+//   2) Activation - activate an installed OS package.
+//   3) Verification - verify that the Activation was successful.
+//
+// Dual Supervisor Target is supported, where the above process is executed once
+// for each Supervisor.
+//
+// Note that certain platforms may have particular approaches to upgrade the
+// firmware of specific components, eg., power supply units, etc.. In addition,
+// platforms may have processes to apply patches to the running OS. Handling
+// these exceptions introduces extra complexities. For Targets that implement
+// this service, component firmware upgrade or OS patching must be embedded
+// within an OS upgrade.
+service OS {
+  // Install transfers an OS package into the Target. No concurrent Install RPCs
+  // must be allowed to the same Target.
+  //
+  // The OS package file format is platform dependent. The platform must
+  // validate that the OS package that is supplied is valid and bootable. This
+  // should include a hash check against a known good hash. It is recommended
+  // that the hash is embedded in the OS package.
+  //
+  // The Target manages its own persistent storage, and OS installation process.
+  // It stores a set of distinct OS packages, and always proactively frees up
+  // space for incoming new OS packages. It is guaranteed that the Target always
+  // has enough space for a valid incoming OS package. The currently running OS
+  // packages must never be removed.
+  //
+  // The Install RPC allows the Client to specify the OS package version. If
+  // the Target already has an OS package with the same version then there is no
+  // need to transfer the OS package to the Target. If the Target does not have
+  // an OS package with the same version, then the OS package is copied.
+  //
+  // Scenario 1 - When the Target already has the OS package:
+  //
+  //         Client :--------------|--------------> Target
+  //              TransferRequest -->
+  //                              <-- [Validated|InstallError]
+  //
+  //
+  // Scenario 2 - When the Target does not have the OS package:
+  //
+  //         Client :--------------|--------------> Target
+  //              TransferRequest -->
+  //                              <-- [TransferReady|InstallError]
+  //            transfer_content  -->
+  //                              ...
+  //                              <-- [TransferProgress|InstallError]
+  //                              ...
+  //                  TransferEnd -->
+  //                              <-- [Validated|InstallError]
+  //
+  // On a dual Supervisor Target, only the Active Supervisor runs this gNOI
+  // Service. The Install RPC applies to the Active Supervisor unless
+  // InstallRequest->TransferRequest->standby_supervisor is set, in which case
+  // it applies to the Standby Supervisor. One Install RPC is required for each
+  // Supervisor.
```

Comment:
And it can be installed on the standby supervisor first in scenario 4 below?

Reply:
Correct. I have updated the text to reflect that.